src/kaleidoscope/plugin/LEDMode.h

```diff
@@ -0,0 +1,148 @@
+* override the mode's colors, the LED mode is the baseline.
+*
+* Most of its functionality is called via @ref LEDControl, with only a few
+* public methods.
+*
+* A LED mode **must** implement at least one of @ref onActivate or @ref
+* update, and possibly @ref refreshAt too.
+*/
+class LEDMode : public kaleidoscope::Plugin,
+  public LEDModeInterface {
+  friend class LEDControl;
+  friend class kaleidoscope::internal::LEDModeManager;
+ protected:
+  // These methods should only be called by LEDControl.
+
+  /** One-time setup, called at keyboard boot.
+   *
+   * Any hooks that need registering, any one-time setup that needs to be
+   * performed, shall be done here. This is purely for preparation purposes, the
+   * LEDs should not be touched yet at this time.
+   */
+  virtual void setup(void) {}
+
+  /** Function to call whenever the mode is activated.
+   *
+   * Like @ref setup, this method need not touch LEDs, @ref update will be
+   * called right after it. The purpose of this callback is to allow a plugin to
+   * do some preparation whenever it is activated, instead of only on boot, or
+   * always at each cycle.
+   *
+   * However, unlike @ref setup, this method can change LED colors, if so
+   * desired. Either to provide an initial state, or a static color set. In the
+   * latter case, consider implementing @ref refreshAt too, because other
+   * plugins may override some of the colors set at activation time, and @ref
+   * refreshAt can be used to restore them when needed.
+   *
+   * Before the callback runs, LEDs will be blanked.
+   */
+  virtual void onActivate(void) {}
+
+  /** Update the LEDs once per cycle.
+   *
+   * Usually the brains of the plugin, which updates the LEDs each cycle. It is
+   * called after the matrix has been scanned, once per cycle.
+   */
+  virtual void update(void) {}
+
+  /** Refresh the color of a given key.
+   *
+   * If we have another plugin that overrides colors set by the active LED mode
+   * (either at @onActivate time, or via @ref update), if that plugin wants to
+   * restore whatever color the mode would set the key color to, this is the
+   * method it will call.
+   *
+   * @param row is the row coordinate of the key to refresh the color of.
+   * @param col is the column coordinate of the key to refresh the color of.
+   */
+  virtual void refreshAt(byte row, byte col) {}
+
+ public:
+
+  /** Plugin initialization.
+   *
+   * Called via `Kaleidoscope.use()`, registers the LED mode, and does the
+   * necessary initialization steps. Calls @ref setup at the end.
+   */
+  kaleidoscope::EventHandlerResult onSetup() {
+    setup();
+
+    return EventHandlerResult::OK;
+  }
+};
+
+} // end namespace plugin
+} // end namespace kaleidoscope
+
+// Invoke this macro in the class definition of a plugin that requires
+// access to its associated transient LED mode.
+//
+#define ACCESS_THIS_LED_MODE                                                   \
```

Comment:
I think we can give this a better name, but I'm not sure what that name is. 

@algernon?

Reply:
I really did not like the magic macro `ACCESS_THIS_LED_MODE` either. Not only because of its name but because of its mere existence. I really would have liked to implement that using inheritance of an auxiliary interface class but I could not because Kaleidoscope at the point where I wrote the code for this PR lacked the ability to check if a class implements a method. 

Now, through the recent enhancement of the event handler signature checks, I sneaked in this ability. By means of an additional commit this PR, I refactored a bit and added a general traits class setup macro `DEFINE_HAS_MEMBER_TRAITS` in a header `has_method.h`. This is now used to check if a LED mode implements the method `registerLEDModeActivated`. This method is inherited from an interface class `AccessTransientLEDMode` that replaces magic macro `ACCESS_THIS_LED_MODE`.

Reply:
BTW, `AccessTransientLEDMode` lives in header `LEDMode.h`.